compiler-user-guide/src/developing-hardware/primitives.md

```diff
@@ -0,0 +1,397 @@
+Because we are creating a _declaration_ primitive the _kind_ must be set to `Declaration`.
+Instead of discussing what the individual template holes mean in the above context, we will instead just give a general listing of the available template holes:
+
+* `~RESULT`: Signal to which the result of a primitive must be assigned
+  to. NB: Only used in a _declaration_ primitive.
+* `~ARG[N]`: `(N+1)`'th argument to the function.
+* `~CONST[N]`: `(N+1)`'th argument to the function. Like `~ARG`, but Clash will try to reduce this to a literal, even if it would otherwise consider it too expensive.
+  And if Clash fails to reduce this argument to a literal it will produce an error.
+* `~LIT[N]`: `(N+1)`'th argument to the function. Like `~CONST~` but values are   rendered as a bare literals, without any size or type annotations.
+  This only works for numeric types, and not for BitVector.
+* `~TYP[N]`: VHDL type of the `(N+1)`'th argument.
+* `~TYPO`: VHDL type of the result.
+* `~TYPM[N]`: VHDL type_name_ of the `(N+1)`'th argument; used in _type qualification_.
+* `~TYPMO`: VHDL type_name_ of the result; used in _type qualification_.
+* `~ERROR[N]`: Error value for the VHDL type of the `(N+1)`'th argument.
+* `~ERRORO`: Error value for the VHDL type of the result.
+* `~GENSYM[<NAME>][N]`: Create a unique name, trying to stay as close to
+  the given `<NAME>` as possible. This unique symbol can be referred to in
+  other places using `~SYM[N]`.
+* `~SYM[N]`: a reference to the unique symbol created by `~GENSYM[<NAME>][N]`.
+* `~SIGD[<HOLE>][N]`: Create a signal declaration, using `<HOLE>` as the name
+  of the signal, and the type of the `(N+1)`'th argument.
+* `~SIGDO[<HOLE>]`: Create a signal declaration, using `<HOLE>` as the name
+  of the signal, and the type of the result.
+* `~TYPEL[<HOLE>]`: The element type of the vector type represented by `<HOLE>`.
+  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~COMPNAME`: The name of the component in which the primitive is instantiated.
+* `~LENGTH[<HOLE>]`: The vector length of the type represented by `<HOLE>`.
+* `~DEPTH[<HOLE>]`: The tree depth of the type represented by `<HOLE>`.
+  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~SIZE[<HOLE>]`: The number of bits needed to encode the type represented by `<HOLE>`.
+  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
+* `~IF <CONDITION> ~THEN <THEN> ~ELSE <ELSE> ~FI`: renders the `<ELSE>`
+  part when `<CONDITION>` evaluates to _0_, and renders the `<THEN>` in all
+  other cases.
+  Valid `<CONDITION>`s are `~LENGTH[<HOLE>]`, `~SIZE[<HOLE>]`, `~CMPLE[<HOLE1>][<HOLE2>]`, `~DEPTH[<HOLE>]`, `~VIVADO`, `~IW64`, `~ISLIT[N]`, `~ISVAR[N]`, `~ISACTIVEENABLE[N]`, `~ISSYNC[N]`, and `~AND[<HOLE1>,<HOLE2>,..]`.
+* `~VIVADO`: _1_ when Clash compiler is invoked with the `-fclash-xilinx` or
+  `-fclash-vivado` flag.
```

Comment:
```suggestion
* `~VIVADO`: _1_ when Clash compiler is invoked with the `-fclash-hdlsyn Vivado` (or `Xilinx` or `ISE`) flag.
```

If we're now also fixing bugs in the same commit, as you've done with some earlier details, I have another one!